File: arcgis_map_android/android/src/main/kotlin/esri/arcgis/flutter_plugin/util/GraphicsParser.kt
package esri.arcgis.flutter_plugin.util

import com.esri.arcgisruntime.geometry.PointCollection
import com.esri.arcgisruntime.geometry.Polygon
import com.esri.arcgisruntime.geometry.Polyline
import com.esri.arcgisruntime.mapping.view.Graphic
import com.esri.arcgisruntime.symbology.PictureMarkerSymbol
import com.esri.arcgisruntime.symbology.SimpleFillSymbol
import com.esri.arcgisruntime.symbology.SimpleLineSymbol
import com.esri.arcgisruntime.symbology.SimpleMarkerSymbol
import com.esri.arcgisruntime.symbology.Symbol
import esri.arcgis.flutter_plugin.model.LatLng
import esri.arcgis.flutter_plugin.model.symbol.PictureMarkerSymbolPayload
import esri.arcgis.flutter_plugin.model.symbol.SimpleFillSymbolPayload
import esri.arcgis.flutter_plugin.model.symbol.SimpleLineSymbolPayload
import esri.arcgis.flutter_plugin.model.symbol.SimpleMarkerSymbolPayload
import esri.arcgis.flutter_plugin.model.toAGSPoint
import esri.arcgis.flutter_plugin.parseToClass

class GraphicsParser {

    companion object {
        fun parse(map: Map<String, Any>): List<Graphic> {

            val graphics = when (val type = map["type"] as String) {
                "point" -> parsePoint(map)
                "polygon" -> parsePolygon(map)
                "polyline" -> parsePolyline(map)
                else -> throw  Exception("No type for $type")
            }

            val attributes = map["attributes"] as? Map<String, String>

            if (attributes != null) {
                return graphics.map {
                    attributes.forEach { (key, value) ->
                        it.attributes[key] = value
                    }
                    it
                }
            }

            return graphics
        }

        private fun parsePoint(map: Map<String, Any>): List<Graphic> {
            val point = (map["point"] as Map<String, Any>).parseToClass<LatLng>()

            val pointGraphic = Graphic().apply {
                geometry = point.toAGSPoint()
                symbol = parseSymbol(map)
            }

            return listOf(pointGraphic)
        }

        private fun parsePolyline(map: Map<String, Any>): List<Graphic> {
            return listOf()
        }

        private fun parsePolygon(map: Map<String, Any>): List<Graphic> {
            return listOf()
        }

        private fun parseSymbol(map: Map<String, Any>): Symbol {
            val symbolMap = map["symbol"] as Map<String, Any>

            val symbol = when (val type = symbolMap["type"]) {
                "simple-marker" -> parseSimpleMarkerSymbol(symbolMap)
                "picture-marker" -> parsePictureMarkerSymbol(symbolMap)
                "simple-fill" -> parseSimpleFillSymbol(symbolMap)
                "simple-line" -> parseSimpleLineSymbol(symbolMap)
                else -> throw Exception("No type for $type")
            }

            return symbol
        }

        private fun parseSimpleMarkerSymbol(map: Map<String, Any>): Symbol {
            val payload = map.parseToClass<SimpleMarkerSymbolPayload>()

            return SimpleMarkerSymbol().apply {
                color = payload.color.toHexInt()
                size = payload.size.toFloat()
                outline = SimpleLineSymbol().apply {
                    style = SimpleLineSymbol.Style.SOLID
                    color = payload.outlineColor.toHexInt()
                    width = payload.outlineWidth.toFloat()
                }
            }
        }

        private fun parsePictureMarkerSymbol(map: Map<String, Any>): Symbol {
            val payload = map.parseToClass<PictureMarkerSymbolPayload>()

            return PictureMarkerSymbol(payload.uri).apply {
                width = payload.width.toFloat()
                height = payload.height.toFloat()
                offsetX = payload.xOffset.toFloat()
                offsetY = payload.yOffset.toFloat()
            }
        }

        private fun parseSimpleFillSymbol(map: Map<String, Any>): Symbol {
            val payload = map.parseToClass<SimpleFillSymbolPayload>()

            return SimpleFillSymbol().apply {
                color = payload.fillColor.toHexInt()
                outline = SimpleLineSymbol().apply {
                    style = SimpleLineSymbol.Style.SOLID
                    color = payload.outlineColor.toHexInt()
                    width = payload.outlineWidth.toFloat()
                }
            }
        }

        private fun parseSimpleLineSymbol(map: Map<String, Any>): Symbol {
            val payload = map.parseToClass<SimpleLineSymbolPayload>()

            return SimpleLineSymbol().apply {
                if (payload.color != null) color = payload.color.toHexInt()
                markerStyle = payload.marker?.style
                markerPlacement = payload.marker?.placement
                style = payload.style
                width = payload.width.toFloat()
            }

        }
    }

}

File: arcgis_map_sdk_ios/ios/arcgis_map_sdk_ios/Sources/arcgis_map_sdk_ios/GraphicsParser.swift
//
//  GraphicsParser.swift
//  arcgis_map_sdk_ios
//
//  Created by Julian Bissekkou on 09.12.22.
//

import Foundation
import ArcGIS
import Flutter

class GraphicsParser {
    let registrar: FlutterPluginRegistrar

    init(registrar: FlutterPluginRegistrar) {
        self.registrar = registrar
    }

    func parse(dictionary: Dictionary<String, Any>) throws -> [Graphic] {
        let type = dictionary["type"] as! String

        let newGraphics: [Graphic]
        switch (type) {
        case "point":
            newGraphics = try! parsePoint(dictionary)
        case "polygon":
            newGraphics = try! parsePolygon(dictionary)
        case "polyline":
            newGraphics = try! parsePolyline(dictionary)
        default:
            throw ParseException(message: "Unknown graphic type: \(type)")
        }

        // Apply attributes to each graphic, if present
        if let attributes = dictionary["attributes"] as? [String: Any] {
            newGraphics.forEach { graphic in
                for (key, value) in attributes {
                    graphic.setAttributeValue(value, forKey: key)
                }
            }
        }
        
        return newGraphics
    }

    private func parsePoint(_ dictionary: [String: Any]) throws -> [Graphic] {
        let point: LatLng = try! JsonUtil.objectOfJson(dictionary["point"] as! Dictionary<String, Any>)

        let graphic = Graphic()
        graphic.geometry = point.toAGSPoint()
        graphic.symbol = try! parseSymbol(dictionary["symbol"] as! Dictionary<String, Any>)

        return [graphic]
    }

    private func parsePolyline(_ dictionary: [String: Any]) throws -> [Graphic] {
        let payload: PathPayload = try! JsonUtil.objectOfJson(dictionary)

        return try payload.paths.map { coordinates in
            let points = try coordinates.map { array in
                if(array.count < 2) {
                    throw ParseException(message: "Size of coordinates need to be at least 2. Got \(array)")
                }
                if(array.count > 2) {
                    return Point(x: array[0], y: array[1], z: array[2], spatialReference: .wgs84)
                }
                return Point(x: array[0], y: array[1], spatialReference: .wgs84)
            }

            let graphic = Graphic()
            graphic.geometry = Polyline(points: points)
            graphic.symbol = try! parseSymbol(dictionary["symbol"] as! Dictionary<String, Any>)

            return graphic
        }
    }

    private func parsePolygon(_ dictionary: [String: Any]) throws -> [Graphic] {
        let payload: PolygonPayload = try! JsonUtil.objectOfJson(dictionary)

        return payload.rings.map { ring in
            let graphic = Graphic()
            let points = ring.map { coordinate in
                Point(x: coordinate[0], y: coordinate[1], spatialReference: .wgs84)
            }
            graphic.geometry = Polygon(points: points)
            graphic.symbol = try! parseSymbol(dictionary["symbol"] as! Dictionary<String, Any>)

            return graphic
        }
    }

    // region symbol parsing

    func parseSymbol(_ dictionary: [String: Any]) throws -> Symbol {
        let type = dictionary["type"] as! String;
        switch (type) {
        case "simple-marker":
            return parseSimpleMarkerSymbol(dictionary)
        case "picture-marker":
            return parsePictureMarkerSymbol(dictionary)
        case "simple-fill":
            return parseSimpleFillMarkerSymbol(dictionary)
        case "simple-line":
            return parseSimpleLineSymbol(dictionary)
        default:
            throw ParseException(message: "Unknown symbol type: \(type)")
        }
    }

    private func parseSimpleMarkerSymbol(_ dictionary: [String: Any]) -> Symbol {
        let payload: SimpleMarkerSymbolPayload = try! JsonUtil.objectOfJson(dictionary)

        let symbol = SimpleMarkerSymbol()
        symbol.color = payload.color.toUIColor()
        symbol.size = payload.size
        symbol.outline = SimpleLineSymbol(
                style: .solid,
                color: payload.outlineColor.toUIColor(),
                width: payload.outlineWidth
        )
        return symbol
    }

    private func parseSimpleFillMarkerSymbol(_ dictionary: [String: Any]) -> Symbol {
        let payload: SimpleFillSymbolPayload = try! JsonUtil.objectOfJson(dictionary)

        let symbol = SimpleFillSymbol()
        symbol.color = payload.fillColor.toUIColor()

        let outline = SimpleLineSymbol()
        outline.width = payload.outlineWidth
        outline.color = payload.outlineColor.toUIColor()
        symbol.outline = outline

        return symbol
    }

    private func parsePictureMarkerSymbol(_ dictionary: [String: Any]) -> Symbol {
        let payload: PictureMarkerSymbolPayload = try! JsonUtil.objectOfJson(dictionary)

        if(!payload.assetUri.isWebUrl()) {
            let uiImage = getFlutterUiImage(payload.assetUri)
            let symbol = PictureMarkerSymbol(image: uiImage!)
            symbol.width = payload.width
            symbol.height = payload.height
            symbol.offsetX = payload.xOffset
            symbol.offsetY = payload.yOffset
            return symbol
        }

        let symbol = PictureMarkerSymbol(url: URL(string: payload.assetUri)!)
        symbol.width = payload.width
        symbol.height = payload.height
        symbol.offsetX = payload.xOffset
        symbol.offsetY = payload.yOffset

        return symbol
    }

    private func getFlutterUiImage(_ fileName: String) -> UIImage? {
            let key = registrar.lookupKey(forAsset: fileName)
        let path = Bundle.main.path(forResource: key, ofType: nil)
        return UIImage(named: path!)
    }

    private func parseSimpleLineSymbol(_ dictionary: [String: Any]) -> Symbol {
        let payload: SimpleLineSymbolPayload = try! JsonUtil.objectOfJson(dictionary)
        let symbol = SimpleLineSymbol()
        
        if let color = payload.color {
            symbol.color = color.toUIColor()
        }
        if let marker = payload.marker {
            symbol.markerStyle = marker.style.toAGSStyle()
            symbol.markerPlacement = marker.placement.toAGSStyle()
        }

        symbol.style = payload.style.toAGSStyle()
        symbol.width = payload.width

        return symbol
    }

    // endregion
}

private struct PathPayload: Codable {
    let paths: [[[Double]]]
}

private struct PolygonPayload: Codable {
    let rings: [[[Double]]]
}

extension String {
    func isWebUrl()-> Bool {
        return starts(with: "https://") || starts(with: "http://")
    }
}
